pubsub/pubsub.go

```diff
@@ -341,26 +339,74 @@ const (
 	// by another process receiving from the same subscription.
 	desiredQueueDuration = 2 * time.Second
 
-	// The factor by which old points decay when a new point is added to the moving
-	// average. The larger this number, the more weight will be given to the newest
-	// point in preference to older ones.
-	decay = 0.05
+	// The initial # of messages to request via ReceiveBatch.
+	initialBatchSize = 1
+
+	// The factor by which old batch sizes decay when a new value is added to the
+	// running value. The larger this number, the more weight will be given to the
+	// newest value in preference to older ones.
+	//
+	// The delta based on a single value is capped by the constants below.
+	decay = 0.5
+
+	// The maximum growth factor in a single jump. Higher values mean that the
+	// batch size can increase more aggressively. For example, 2.0 means that the
+	// batch size will at most double from one ReceiveBatch call to the next.
+	maxGrowthFactor = 2.0
+
+	// Similarly, the maximum shrink factor. Lower values mean that the batch size
+	// can shrink more aggressively. For example; 0.75 means that the batch size
+	// will at most shrink to 75% of what it was before. Note that values less
+	// than (1-decay) will have no effect because the running value can't change
+	// by more than that.
+	maxShrinkFactor = 0.75
+
+	// The maximum batch size to request.
+	// TODO(rvangent): Increase this.
+	maxBatchSize = 1000
 )
 
-// Add message processing time d to the weighted moving average.
-func (s *Subscription) addProcessingTime(d time.Duration) {
-	// Ensure d is non-zero; on some platforms, the granularity of time.Time
-	// isn't small enough to capture very fast message processing times.
-	if d == 0 {
-		d = 1 * time.Nanosecond
+// updateBatchSize updates the # of messages to request in ReceiveBatch based
+// on the previous batch size and rate of messages being pulled from the queue.
+//
+// It returns the number of messages to request in this ReceiveBatch call.
+//
+// s.mu must be held.
+func (s *Subscription) updateBatchSize() int {
+	if s.lastBatchRecv.IsZero() {
+		// First call, no data yet.
+		s.runningBatchSize = initialBatchSize
+		return initialBatchSize
 	}
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	if s.avgProcessTime == 0 {
-		s.avgProcessTime = d.Seconds()
+	if !s.dynamicBatchSizes {
```

Comment:
Thought: maybe a simpler and more flexible workaround for testing would be a `fixedBatchSize` field. Then at the top of this function:
```
if s.fixedBatchedSize != 0 {
  return s.fixedBatchSize
}
```

Reply:
SGTM, will do separately.